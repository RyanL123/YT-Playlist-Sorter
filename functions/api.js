// var express = require("express");
// var router = express.Router();
// require("dotenv").config();
// const fetch = require("node-fetch");

// router.get("/playlist", async (req, res) => {
//     const id = req.query.id;
//     const data = await fetch(
//         `https://www.googleapis.com/youtube/v3/playlistItems?part=contentDetails%2Csnippet&maxResults=50&playlistId=${id}&key=${process.env.YOUTUBE_API_KEY}`
//     ).then((res) => {
//         return res.json();
//     });
//     return res.json(data);
// });

// router.get("/video", async (req, res) => {
//     const id = req.query.id;
//     const data = await fetch(
//         `https://www.googleapis.com/youtube/v3/videos?part=statistics%2Csnippet&id=${id}&key=${process.env.YOUTUBE_API_KEY}`
//     ).then((res) => {
//         return res.json();
//     });
//     return res.json(data);
// });

// module.exports = router;
