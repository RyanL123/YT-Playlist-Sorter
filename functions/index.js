const functions = require("firebase-functions");
const cors = require("cors")({ origin: true });
require("dotenv").config();
const fetch = require("node-fetch");

exports.playlist = functions.https.onCall(async (data, context) => {
  const id = data.id;
  const token = data.token || ""; // make sure token is not undefined
  return await fetch(
    `https://www.googleapis.com/youtube/v3/playlistItems?part=contentDetails%2Csnippet&maxResults=50&playlistId=${id}&pageToken=${token}&key=${process.env.YOUTUBE_API_KEY}`,
  ).then((res) => {
    return res.json();
  });
});

exports.video = functions.https.onCall(async (data, context) => {
  const id = data.id;
  return await fetch(
    `https://www.googleapis.com/youtube/v3/videos?part=statistics%2Csnippet%2CcontentDetails&id=${id}&key=${process.env.YOUTUBE_API_KEY}`,
  ).then((res) => {
    return res.json();
  });
});
