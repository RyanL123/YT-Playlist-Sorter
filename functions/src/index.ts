import { youtube } from "@googleapis/youtube";
import * as logger from "firebase-functions/logger";
import { onCall } from "firebase-functions/v2/https";

const youtubeClient = youtube({
  auth: process.env.YOUTUBE_API_KEY,
  version: "v3",
});

export const getVideosFromPlaylist = onCall(async (request) => {
  const playlistId = request.data.id;
  const playlist = await youtubeClient.playlistItems.list({
    playlistId: playlistId,
    part: ["contentDetails", "snippet"],
    maxResults: 50,
  });

  const dataItems = playlist?.data?.items ?? [];

  let videos = await Promise.all(
    dataItems.map((playlistItem) => {
      const videoId = playlistItem.snippet?.resourceId?.videoId ?? "";
      return youtubeClient.videos.list({
        part: ["statistics", "snippet", "contentDetails"],
        id: [videoId],
      });
    }),
  );

  let filteredVideos = videos
    .map((video) => video?.data?.items?.[0])
    .filter((video) => video != undefined);

  logger.debug(filteredVideos);
  return { videos: filteredVideos };
});
