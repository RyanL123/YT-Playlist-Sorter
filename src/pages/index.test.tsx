import { fireEvent, render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import React from "react";
import { BrowserRouter } from "react-router-dom";
import renderer from "react-test-renderer";
import Index from "./index";
import "@testing-library/jest-dom/extend-expect";
import { getFunctions, httpsCallable } from "firebase/functions";
import type {
  PlaylistItemListResponse,
  VideoListResponse,
  VideoMetadata,
} from "../types";

const mockVideosById: Record<string, VideoMetadata> = {
  abc: {
    id: "abc",
    statistics: {
      viewCount: "1",
      likeCount: "1",
    },
    snippet: {
      publishedAt: "2000/01/01",
      channelTitle: "A",
      title: "Video1",
      thumbnails: {
        medium: {
          url: "abc.com",
        },
      },
    },
    contentDetails: {
      duration: "PT10S",
    },
  },
  def: {
    id: "abc",
    statistics: {
      viewCount: "1",
      likeCount: "1",
    },
    snippet: {
      publishedAt: "2000/01/01",
      channelTitle: "A",
      title: "Video2",
      thumbnails: {
        medium: {
          url: "abc.com",
        },
      },
    },
    contentDetails: {
      duration: "PT10S",
    },
  },
};

const mockPlaylistsById: Record<string, PlaylistItemListResponse> = {
  playlist1: {
    kind: "youtube#playlistItemListResponse",
    nextPageToken: "",
    prevPageToken: "",
    pageInfo: {
      totalResults: 1,
      resultsPerPage: 1,
    },
    items: [
      {
        kind: "youtube#playlistItem",
        snippet: {
          resourceId: {
            videoId: "abc",
          },
        },
      },
    ],
  },
  playlist2: {
    kind: "youtube#playlistItemListResponse",
    nextPageToken: "",
    prevPageToken: "",
    pageInfo: {
      totalResults: 1,
      resultsPerPage: 1,
    },
    items: [
      {
        kind: "youtube#playlistItem",
        snippet: {
          resourceId: {
            videoId: "def",
          },
        },
      },
    ],
  },
};

const mockVideoFunction = ({ id }): { data: VideoListResponse } => {
  return {
    data: {
      kind: "youtube#videoListResponse",
      nextPageToken: "",
      prevPageToken: "",
      pageInfo: {
        totalResults: 1,
        resultsPerPage: 1,
      },
      items: [mockVideosById[id]],
    },
  };
};

const mockPlaylistFunction = ({ id }): { data: PlaylistItemListResponse } => {
  return { data: mockPlaylistsById[id] };
};

jest.mock("firebase/functions", () => {
  return {
    __esModule: true,
    ...jest.requireActual("firebase/functions"),
    httpsCallable: (_, name) => {
      if (name === "video") {
        return mockVideoFunction;
      } else if (name === "playlist") {
        return mockPlaylistFunction;
      }
    },
  };
});

it("renders correctly", async () => {
  const component = renderer.create(
    <BrowserRouter>
      <Index />
    </BrowserRouter>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();

  const user = userEvent.setup();

  render(
    <BrowserRouter>
      <Index />
    </BrowserRouter>,
  );

  const playlistIdInput = screen.getByTestId("playlistIdInput");
  await user.click(playlistIdInput);
  await user.keyboard("playlist1");

  const form = screen.getByTestId("searchPanelForm");
  fireEvent.submit(form);

  await waitFor(
    async () => {
      expect(
        await screen.findAllByText("Video1", { exact: false }),
      ).toBeDefined();
    },
    {
      timeout: 5000,
    },
  );

  await user.clear(playlistIdInput);
  await user.click(playlistIdInput);
  await user.keyboard("playlist2");

  fireEvent.submit(form);

  await waitFor(
    async () => {
      expect(
        await screen.findAllByText("Video2", { exact: false }),
      ).toBeDefined();
    },
    {
      timeout: 5000,
    },
  );
}, 20000);
