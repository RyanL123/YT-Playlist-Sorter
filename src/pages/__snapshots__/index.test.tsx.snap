// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
Array [
  <div
    className="MuiBox-root css-x6sufn"
  >
    <form
      data-testid="searchPanelForm"
      onSubmit={[Function]}
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <div
        className="MuiBox-root css-olp8is"
      >
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-sizeMedium MuiInputLabel-outlined css-14s5rfu-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={false}
            htmlFor=":r0:"
            id=":r0:-label"
          >
            Playlist ID or Link
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <input
              aria-invalid={false}
              autoFocus={false}
              className="MuiInputBase-input MuiOutlinedInput-input css-1t8l2tu-MuiInputBase-input-MuiOutlinedInput-input"
              data-testid="playlistIdInput"
              disabled={false}
              id=":r0:"
              name="playlistID"
              onAnimationStart={[Function]}
              onBlur={[Function]}
              onChange={[Function]}
              onFocus={[Function]}
              required={false}
              type="text"
              value=""
            />
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-yjsfm1"
              >
                <span>
                  Playlist ID or Link
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          className="MuiFormControl-root MuiFormControl-fullWidth MuiTextField-root css-wb57ya-MuiFormControl-root-MuiTextField-root"
        >
          <label
            className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined MuiFormLabel-colorPrimary MuiFormLabel-filled MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-sizeMedium MuiInputLabel-outlined css-1jy569b-MuiFormLabel-root-MuiInputLabel-root"
            data-shrink={true}
            htmlFor=":r1:"
            id=":r1:-label"
          >
            Order
          </label>
          <div
            className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary MuiInputBase-fullWidth MuiInputBase-formControl css-md26zr-MuiInputBase-root-MuiOutlinedInput-root"
            onClick={[Function]}
          >
            <div
              aria-controls=":r2:"
              aria-expanded="false"
              aria-haspopup="listbox"
              aria-labelledby=":r1:-label :r1:"
              className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
              id=":r1:"
              onBlur={[Function]}
              onFocus={[Function]}
              onKeyDown={[Function]}
              onMouseDown={[Function]}
              role="combobox"
              tabIndex={0}
            >
              Views Descending
            </div>
            <input
              aria-hidden={true}
              aria-invalid={false}
              autoFocus={false}
              className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
              disabled={false}
              name="order"
              onAnimationStart={[Function]}
              onChange={[Function]}
              required={false}
              tabIndex={-1}
              value="Views Descending"
            />
            <svg
              aria-hidden={true}
              className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
              data-testid="ArrowDropDownIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M7 10l5 5 5-5z"
              />
            </svg>
            <fieldset
              aria-hidden={true}
              className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
            >
              <legend
                className="css-14lo706"
              >
                <span>
                  Order
                </span>
              </legend>
            </fieldset>
          </div>
        </div>
        <div
          className="MuiBox-root css-2bwllc"
        >
          <button
            className="MuiButtonBase-root MuiButton-root MuiLoadingButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiLoadingButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1pkfx9i-MuiButtonBase-root-MuiButton-root-MuiLoadingButton-root"
            disabled={false}
            id=":r3:"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
            type="submit"
          >
            Search
            <span
              className="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="SearchIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M15.5 14h-.79l-.28-.27C15.41 12.59 16 11.11 16 9.5 16 5.91 13.09 3 9.5 3S3 5.91 3 9.5 5.91 16 9.5 16c1.61 0 3.09-.59 4.23-1.57l.27.28v.79l5 4.99L20.49 19zm-6 0C7.01 14 5 11.99 5 9.5S7.01 5 9.5 5 14 7.01 14 9.5 11.99 14 9.5 14"
                />
              </svg>
            </span>
          </button>
          <a
            className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-colorPrimary css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
            href="/help"
            onBlur={[Function]}
            onClick={[Function]}
            onContextMenu={[Function]}
            onDragLeave={[Function]}
            onFocus={[Function]}
            onKeyDown={[Function]}
            onKeyUp={[Function]}
            onMouseDown={[Function]}
            onMouseLeave={[Function]}
            onMouseUp={[Function]}
            onTouchEnd={[Function]}
            onTouchMove={[Function]}
            onTouchStart={[Function]}
            tabIndex={0}
          >
            Help
            <span
              className="MuiButton-icon MuiButton-endIcon MuiButton-iconSizeMedium css-9tj150-MuiButton-endIcon"
            >
              <svg
                aria-hidden={true}
                className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="InfoOutlinedIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M11 7h2v2h-2zm0 4h2v6h-2zm1-9C6.48 2 2 6.48 2 12s4.48 10 10 10 10-4.48 10-10S17.52 2 12 2m0 18c-4.41 0-8-3.59-8-8s3.59-8 8-8 8 3.59 8 8-3.59 8-8 8"
                />
              </svg>
            </span>
          </a>
        </div>
      </div>
    </form>
  </div>,
  <div
    className="MuiBox-root css-113h623"
  />,
]
`;
